stdlib/hashtbl.ml

```diff
@@ -287,209 +385,100 @@ module type SeededHashedType =
-          | Cons{key=k2; data=d2; next=next2} ->
-              if H.equal key k2 then d2 else
-              match next2 with
-              | Empty -> raise Not_found
-              | Cons{key=k3; data=d3; next=next3} ->
-                  if H.equal key k3 then d3 else find_rec key next3
-
-    let rec find_rec_opt key = function
-      | Empty ->
-          None
-      | Cons{key=k; data; next} ->
-          if H.equal key k then Some data else find_rec_opt key next
-
-    let find_opt h key =
-      match h.data.(key_index h key) with
-      | Empty -> None
-      | Cons{key=k1; data=d1; next=next1} ->
-          if H.equal key k1 then Some d1 else
-          match next1 with
-          | Empty -> None
-          | Cons{key=k2; data=d2; next=next2} ->
-              if H.equal key k2 then Some d2 else
-              match next2 with
-              | Empty -> None
-              | Cons{key=k3; data=d3; next=next3} ->
-                  if H.equal key k3 then Some d3 else find_rec_opt key next3
-
-    let find_all h key =
-      let[@tail_mod_cons] rec find_in_bucket = function
-      | Empty ->
-          []
-      | Cons{key=k; data=d; next} ->
-          if H.equal k key
-          then d :: find_in_bucket next
-          else find_in_bucket next in
-      find_in_bucket h.data.(key_index h key)
-
-    let rec replace_bucket key data = function
-      | Empty ->
-          true
-      | Cons ({key=k; next} as slot) ->
-          if H.equal k key
-          then (slot.key <- key; slot.data <- data; false)
-          else replace_bucket key data next
-
-    let replace h key data =
-      let i = key_index h key in
-      let l = h.data.(i) in
-      if replace_bucket key data l then begin
-        h.data.(i) <- Cons{key; data; next=l};
-        h.size <- h.size + 1;
-        if h.size > Array.length h.data lsl 1 then resize key_index h
-      end
-
-    let rec mem_in_bucket key = function
-      | Empty ->
-          false
-      | Cons{key=k; next} ->
-          H.equal k key || mem_in_bucket key next
-
-    let mem h key =
-      mem_in_bucket key h.data.(key_index h key)
-
-    let add_seq tbl i =
-      Seq.iter (fun (k,v) -> add tbl k v) i
-
-    let replace_seq tbl i =
-      Seq.iter (fun (k,v) -> replace tbl k v) i
-
-    let of_seq i =
-      let tbl = create 16 in
-      replace_seq tbl i;
-      tbl
-
-    let iter = iter
-    let filter_map_inplace = filter_map_inplace
-    let fold = fold
-    let length = length
-    let stats = stats
-    let to_seq = to_seq
-    let to_seq_keys = to_seq_keys
-    let to_seq_values = to_seq_values
-  end
-
-module Make(H: HashedType): (S with type key = H.t) =
-  struct
-    include MakeSeeded(struct
-        type t = H.t
-        let equal = H.equal
-        let seeded_hash (_seed: int) x = H.hash x
-      end)
-    let create sz = create ~random:false sz
-    let of_seq i =
-      let tbl = create 16 in
-      replace_seq tbl i;
-      tbl
-  end
+module type S = sig
```

Comment:
I would recommend to limit the scope of the PR to the refactorings that @andrepd initially planned, and consider further improvements in follow-up PRs, instead of going full refactor mode right away. If we have to discuss performance impact at the same time, the less unrelated changes the better.